File: SASMopubNativeAdCustomEvent.m
//
//  SASMopubNativeAdCustomEvent.m
//  SmartAdServer
//
//  Created by Thomas Geley on 26/12/2016.
//  Copyright © 2019 Smart AdServer. All rights reserved.
//

#import "SASMopubNativeAdCustomEvent.h"
#import "SASMopubNativeAdAdapter.h"
#import "SASMopubUtils.h"

NS_ASSUME_NONNULL_BEGIN

@interface SASMopubNativeAdCustomEvent ()

@property (nonatomic, strong, nullable) SASNativeAdManager *sasNativeAdManager;

@end

@implementation SASMopubNativeAdCustomEvent

#pragma mark - Dealloc

- (void)dealloc {
    [self resetNativeAdManager];
}

- (void)resetNativeAdManager {
    self.sasNativeAdManager = nil;
}

#pragma mark - Custom Event Extension

- (void)requestAdWithCustomEventInfo:(NSDictionary *)info {
    
    // Reset native ad manager
    [self resetNativeAdManager];
    
    // Extracting placement from custom event info
    NSError *error = nil;
    SASAdPlacement *adPlacement = [SASMopubUtils adPlacementWithCustomEventInfo:info error:&error];
    
    if (adPlacement == nil) {
        // Failing if custom info are invalid
        [self.delegate nativeCustomEvent:self didFailToLoadAdWithError:error];
        return;
    }
        
    // Creating a native ad manager instance
    self.sasNativeAdManager = [[SASNativeAdManager alloc] initWithPlacement:adPlacement];
    
    // Requesting an ad for the current placement
    [self.sasNativeAdManager requestAd:^(SASNativeAd * _Nullable ad, NSError * _Nullable error) {
        if (ad) {
            [self sasNativeAdDidLoad:ad];
        } else {
            [self sasNativeAdDidFailToLoadWithError:error];
        }
    }];

}

#pragma mark - Ad Loading Methods

- (void)sasNativeAdDidLoad:(SASNativeAd *)nativeAd {
    
    // Adapter initialization
    SASMopubNativeAdAdapter *adAdapter = [[SASMopubNativeAdAdapter alloc] initWithSASNativeAd:nativeAd];
    
    // MoPub ad initialization
    MPNativeAd *interfaceAd = [[MPNativeAd alloc] initWithAdAdapter:adAdapter];
    
    // Precaching images if possible…
    NSMutableArray *imageURLs = [NSMutableArray array];
    
    if (nativeAd.icon.URL) {
        [imageURLs addObject:nativeAd.icon.URL];
    }
    
    if (nativeAd.coverImage.URL) {
        [imageURLs addObject:nativeAd.coverImage.URL];
    }
    
    [super precacheImagesWithURLs:imageURLs completionBlock:^(NSArray *errors) {
        if (errors) {
            MPLogInfo(@"Error(s) when precaching SASNativeAd image(s) : %@", errors);
            [self.delegate nativeCustomEvent:self didFailToLoadAdWithError:MPNativeAdNSErrorForImageDownloadFailure()];
        } else {
            [self.delegate nativeCustomEvent:self didLoadAd:interfaceAd];
        }
    }];
    
}

- (void)sasNativeAdDidFailToLoadWithError:(NSError *)error {
    MPLogInfo(@"Error when loading SASNativeAd : %@", error);
    [self.delegate nativeCustomEvent:self didFailToLoadAdWithError:error];
}

@end

NS_ASSUME_NONNULL_END

File: SASMopubNativeAdCustomEvent.swift
//
//  SASMopubNativeAdCustomEvent.swift
//  SASMopubNativeAdCustomEvent
//
//  Created by Loïc GIRON DIT METAZ on 02/09/2021.
//

import Foundation
import SASDisplayKit
import MoPubSDK

@objc(SASMopubNativeAdCustomEvent)
class SASMopubNativeAdCustomEvent: MPNativeCustomEvent {
    
    private var sasNativeAdManager: SASNativeAdManager? = nil
    
    override func requestAd(withCustomEventInfo info: [AnyHashable : Any]!, adMarkup: String!) {
        do {
            // Extracting placement from custom event info
            let adPlacement = try SASMopubUtils.adPlacement(customEventInfo: info)
            
            // Creating a native ad manager instance
            sasNativeAdManager = SASNativeAdManager(placement: adPlacement)
            
            // Requesting an ad for the current placement
            self.sasNativeAdManager?.requestAd({ [self] nativeAd, error in
                if let nativeAd = nativeAd {
                    sasNativeAdDidLoad(nativeAd: nativeAd)
                } else {
                    sasNativeAdDidFailToLoad(error: error)
                }
            })
        } catch {
            // Failing if custom info are invalid
            delegate?.nativeCustomEvent(self, didFailToLoadAdWithError: error)
        }
    }
    
    private func sasNativeAdDidLoad(nativeAd: SASNativeAd) {
        // Adapter initialization
        let adAdapter = SASMopubNativeAdAdapter(nativeAd: nativeAd)
        
        // MoPub ad initialization
        let interfaceAd = MPNativeAd(adAdapter: adAdapter)
        
        // Precaching images if possible…
        var imageURLs = [URL]()
        if let url = nativeAd.icon?.url {
            imageURLs.append(url)
        }
        if let url = nativeAd.coverImage?.url {
            imageURLs.append(url)
        }
        
        super.precacheImages(withURLs: imageURLs) { [self] errors in
            if errors != nil {
                delegate?.nativeCustomEvent(self, didFailToLoadAdWithError: MPNativeAdNSErrorForImageDownloadFailure())
            } else {
                delegate?.nativeCustomEvent(self, didLoad: interfaceAd)
            }
        }
    }
    
    private func sasNativeAdDidFailToLoad(error: Error?) {
        delegate?.nativeCustomEvent(self, didFailToLoadAdWithError: error)
    }
    
}
